python/ql/src/semmle/python/dataflow/new/internal/Attributes.qll

```diff
@@ -17,6 +17,15 @@ abstract class AttrRef extends Node {
    */
   abstract Node getObject();
 
+  /**
+   * Gets the data flow node corresponding to the object whose attribute named
+   * `attrName` is being read or written.
+   */
+  Node getObject(string attrName) {
```

Comment:
The JavaScript libraries define an [`accesses`](https://github.com/github/codeql/blob/main/javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll#L531) method for the case where you want to link up the nodes in question. Would this work as an alternative?

Reply:
I tried this out now, and while it _would_ work, it doesn't end up being as nice. I added it as

```codeql
  /**
   * Holds if this data flow node accesses attribute named `attrName` on object `object`.
   */
  predicate accesses(Node object, string attrName) {
    this.getObject() = object and getAttributeName() = attrName
  }
```

I found two places where this needed to be replaced:

```diff
@@ -172,13 +172,15 @@ predicate containerStep(DataFlow::CfgNode nodeFrom, DataFlow::Node nodeTo) {
         // dict
         "values", "items", "get", "popitem"
       ] and
-    call.getFunction().(DataFlow::AttrRead).getObject(name) = nodeFrom
+    call.getFunction().(DataFlow::AttrRead).accesses(nodeFrom, name)
   )
   or
   // list.append, set.add
   exists(DataFlow::CallCfgNode call, string name |
     name in ["append", "add"] and
-    call.getFunction().(DataFlow::AttrRead).getObject(name).getPostUpdateNode() = nodeTo and
+    call.getFunction()
+        .(DataFlow::AttrRead)
+        .accesses(any(DataFlow::Node obj | obj.getPostUpdateNode() = nodeTo), name) and
     call.getArg(0) = nodeFrom
   )
 }
```

I don't really mind the first one. I think either looks fine, and I guess my slight bias towards `.getObject` is just based on being used to seeing that one.

I really don't like the second change though, it becomes much more obscure to read (and write) in my opinion. The fact that member-predicates can be chained with `getObject(name).getPostUpdateNode()` is really nice.

So based on this experiment, I would like to go ahead with the `getObject(name)` added in this pr, even though it is not able to align with JS. Does that sound ok?

Reply:
I agree that the second one looks less than ideal, but I think in this case it's because of another shortcoming in our API. The awkward part here is the fact that we have to invert `getPostUpdateNode` with an `any`. If instead we could write it as `.accesses(nodeTo.getPreUpdateNode(), name)`, it would look much nicer. And honestly, I find this much more intuitive than `getObject/1`. (My main objection is that the argument to `getObject/1` seems totally disconnected from its result. After all, `getObject/0` returns the exact same node.)

I am tempted to suggest a third alternative, in the form of a method on `AttrRef` like this:
```ql
DataFlow::Node withName(string name) {
  result = this and
  this.getName() = name
}
```
That would at least make it clear what the role of the otherwise magic argument to `getObject/1` is (at least, to me it reads better with `attr_ref.withName(name).getObject()`). However, I think this is also a mistake, and it's better overall to align on a common API.

Reply:
I guess we could do `.accesses(nodeTo.(DataFlow::PostUpdateNote).getPreUpdateNode(), name)` which is slightly shorter than `.accesses(any(DataFlow::Node obj | obj.getPostUpdateNode() = nodeTo), name)`

Reply:
Here's something that would make your first snippet above a bit cleaner: https://github.com/github/codeql/pull/6079
I think with this, `.accesses` actually looks pretty intuitive and clean.